workers/common/src/main/kotlin/common/env/NetRcManager.kt

```diff
@@ -0,0 +1,114 @@
+ * the set of infrastructure services currently available. It is also registered at the central authenticator used by
+ * ORT Server. Its task is to figure out, based on authentication events, which infrastructure services are currently
+ * referenced. These are then written into the `.netrc` file, if they have the corresponding credentials type. This
+ * should prevent conflicts with multiple services for the same host.
+ */
+internal class NetRcManager(
+    /** The current [WorkerContext]. */
+    private val context: WorkerContext,
+
+    /** The currently known set of infrastructure services. */
+    services: Collection<InfrastructureService>
+) : AuthenticationListener {
+    companion object {
+        private val logger = LoggerFactory.getLogger(NetRcManager::class.java)
+
+        /**
+         * Create a new instance of [NetRcManager] and initialize it with the given [context] and [services].
+         */
+        fun create(context: WorkerContext, services: Collection<InfrastructureService>): NetRcManager =
+            NetRcManager(context, services)
+    }
+
+    /** A map with all relevant services using the service name as key. */
+    private val netRcServices = services.filter { CredentialsType.NETRC_FILE in it.credentialsTypes }
+        .associateBy { it.name }
+
+    /**
+     * Stores the services for which authentication events have been received. These are the services that are
+     * currently referenced from the `.netrc` file. The key is the host for which a service is responsible for.
+     */
+    private val authenticatedServices = mutableMapOf<String, EnvironmentServiceDefinition>()
+
+    /** A mutex to guard access to the authenticated services and updates of the `.netrc` file. */
+    private val mutex = Mutex()
+
+    override fun onAuthentication(authenticationEvent: AuthenticationEvent) {
+        logger.info("Received authentication event for service '${authenticationEvent.serviceName}'.")
+
+        netRcServices[authenticationEvent.serviceName]?.also { service ->
+            runBlocking { updateNetRcServices(service) }
+        }
+    }
+
+    /**
+     * Return a new instance of [ConfigFileBuilder] to be used for creating the `.netrc` file.
+     */
+    internal fun createConfigFileBuilder(): ConfigFileBuilder = ConfigFileBuilder(context)
+
+    /**
+     * Return a new instance of [NetRcGenerator] that is going to be called to generate the `.netrc` file.
+     */
+    internal fun createNetRcGenerator(): NetRcGenerator = NetRcGenerator()
+
+    /**
+     * Perform an update of the `.netrc` file after receiving an authentication event for the given [service].
+     */
+    private suspend fun updateNetRcServices(service: InfrastructureService) {
+        val serviceUri = URI.create(service.url)
+        val serviceHost = serviceUri.host
+
+        val builder = createConfigFileBuilder()
```

Comment:
Instead of the `WorkerContext` this class could just take the `builder` and `generator` as constructor arguments to be more explicit  about what it actually needs. The `WorkerContext` has more functionality (and dependencies) than needed here.
It would also simplify testing a bit as you would not have to mock the two trivial functions above.

Reply:
A builder instance can only be used for a single file generation. Therefore, for every received event, a new builder instance has to be created, and this requires a context.

Reply:
I'm unhappy about that `WorkerContext` god object, the builder actually only needs a secret provider and only uses the helper code from the context to resolve secrets in parallel, but with the context it get dependencies on other stuff like downloading config files.